quickjs-libc.c

```diff
@@ -2890,6 +2890,65 @@ static int my_execvpe(const char *filename, char **argv, char **envp)
     return -1;
 }
 
+/*
+    This method is called from the child process (after forking)
+    and closes all open file descriptors > 2
+ */
+void close_fds() {
+/*
+    Use closefrom if possible
+ */
+#if defined(HAVE_CLOSEFROM)
+    closefrom(3);
+#else
+    int fd_max = sysconf(_SC_OPEN_MAX);
+    uint32_t i;
+#if defined(__linux__)
+    /*
+        Under linux
+
+        - check /proc to find all open file descriptors > 2 and close them
+        - set fd_max = 3 if at least one fd was found in /proc
+
+        This will improve performances on systems where the maximum number
+        of open file descriptor is high (such as in a Docker container).
+    */
+    int pid = getpid();
+    char path[32];
+    struct stat statbuf;
+    snprintf(path, sizeof(path), "/proc/%d/fd", pid);
+    if (stat(path, &statbuf) == 0) {
+        if (S_ISDIR(statbuf.st_mode)) {
+            DIR *dir = opendir(path);
+            if (dir) {
+                struct dirent *subdir;
+                int fd;
+                /*
+                    Close all open file descriptors > 2 and set fd_max to 3
+                */
+                for(;;) {
+                    subdir = readdir(dir);
+                    if (!subdir) {
+                        break;
+                    }
+                    fd = atoi(subdir->d_name);
+                    if (fd > 0) {
+                        fd_max = 3;
+                        if (fd > 2 && fd != dirfd(dir)) {
+                            close(fd);
+                        }
+                    }
+                }
+                closedir(dir);
+            }
+        }
+    }
+#endif
```

Comment:
Since linux has had `closefrom` for a while now, this approach does not seem useful.

Reply:
We still need this approach because closefrom is not available on alpine

Reply:
This configuration setting is indeed system dependant. I have observed these values on some of my systems:
- my laptop running a recent macOS: 256
- Debian linux boxes: 1024 and 65535
- FreeBSD: 28404
- OpenBSD: 512

Modern systems have `closefrom` so we should use that if available and the problem is solved correctly.

On obsolete systems, a best effort approach seems reasonable:
- parsing `/proc/` may actually not be possible as this file system may be disabled.  Furthermore if would only work in linux.
- using `sysconf(_SC_OPEN_MAX)` is reliable but we should limit the loop to a reasonable maximum such as 1024.
- file handles should be open with `O_CLOEXEC` as much as possible.
- it is unlikely more than a few handles will have been opened before `os.exec` is called.
- not closing the handles beyond 1024 on such systems does not have dire consequences.